src/row_codec.rs

```diff
@@ -1,182 +1,230 @@
-        ValueDeletable::Value(data.slice(..value_len))
-    };
-
-    let mut key = BytesMut::with_capacity(key_prefix_len + key_suffix_len);
-    key.extend_from_slice(&first_key[..key_prefix_len]);
-    key.extend_from_slice(&key_suffix);
-
-    Ok(KeyValueDeletable {
-        key: key.into(),
-        value,
-        attributes: RowAttributes {
-            ts: meta.timestamp,
-            expire_ts: meta.expire_ts,
-        },
-    })
-}
+pub(crate) struct SstRowCodecV1 {}
 
-struct RowMetadata {
-    flags: RowFlags,
-    timestamp: Option<i64>,
-    expire_ts: Option<i64>,
-}
+impl SstRowCodecV1 {
+    pub fn new() -> Self {
+        Self {}
+    }
 
-fn decode_meta(row_features: &[RowFeature], data: &mut Bytes) -> Result<RowMetadata, SlateDBError> {
-    let mut meta = RowMetadata {
-        flags: RowFlags::empty(),
-        timestamp: None,
-        expire_ts: None,
-    };
-    for attr in row_features {
-        match attr {
-            RowFeature::Flags => match decode_row_flags(data.get_u8()) {
-                Ok(flags) => {
-                    meta.flags = flags;
-                }
-                Err(e) => return Err(e),
-            },
-            RowFeature::Timestamp => meta.timestamp = Some(data.get_i64()),
-            RowFeature::ExpireAtTs => {
-                let expire_ts = data.get_i64();
-                meta.expire_ts = if expire_ts == NO_EXPIRE_TS {
-                    None
-                } else {
-                    Some(expire_ts)
-                };
+    pub fn encode(&self, output: &mut Vec<u8>, row: &SstRowEntry) {
+        match &row.key {
+            SstRowKey::SuffixOnly { prefix_len, suffix } => {
+                output.put_u16(*prefix_len as u16);
+                output.put_u16(suffix.len() as u16);
+                output.put(suffix.as_ref());
+            }
+            _ => unreachable!("only suffix only key is supported on encode"),
+        }
+
+        let flags = {
+            let mut flags = RowFlags::empty();
+            if row.expire_ts.is_some() {
+                flags |= RowFlags::HAS_EXPIRE_TS;
+            }
+            if row.create_ts.is_some() {
+                flags |= RowFlags::HAS_CREATE_TS;
+            }
+            if row.value.as_option().is_none() {
+                flags |= RowFlags::Tombstone;
             }
+            flags
+        };
+
+        // encode seq & flags
+        output.put_u64(row.seq);
+        output.put_u8(flags.bits());
+        if flags.contains(RowFlags::Tombstone) {
+            return;
         }
+
+        // encode expire & create ts
+        if flags.contains(RowFlags::HAS_EXPIRE_TS) {
+            output.put_i64(row.expire_ts.unwrap_or(0));
+        }
+        if flags.contains(RowFlags::HAS_CREATE_TS) {
+            output.put_i64(row.create_ts.unwrap_or(0));
+        }
+
+        // encode value
+        let val = row
+            .value
+            .as_option()
+            .expect("value is not set with no tombstone");
```

Comment:
do we not want to support `None` as a value and just have value length be `0`? that's one of the benefits of having a specific tombstone flag, supporting None values is a nice feature IMO